pkg/collector/corechecks/system/winregistry/winregistry.go

```diff
@@ -0,0 +1,210 @@
+		return err
+	}
+
+	var conf checkCfg
+	if err := yaml.Unmarshal(data, &conf); err != nil {
+		return err
+	}
+
+	hiveMap := map[string]registry.Key{
+		"HKLM":               registry.LOCAL_MACHINE,
+		"HKEY_LOCAL_MACHINE": registry.LOCAL_MACHINE,
+		"HKU":                registry.USERS,
+		"HKEY_USERS":         registry.USERS,
+		"HKEY_CLASSES_ROOT":  registry.CLASSES_ROOT,
+		"HKCR":               registry.CLASSES_ROOT,
+	}
+
+	for regKey, regKeyConfig := range conf.RegistryKeys {
+		splitKeypath := strings.Split(regKey, "\\")
+		if len(splitKeypath) < 2 {
+			return log.Errorf("the key %s is too short to be a valid key", regKey)
+		}
+
+		if len(regKeyConfig.Name) == 0 {
+			log.Warnf("the key %s does not have a metric name, skipping", regKey)
+			continue
+		}
+
+		if hive, found := hiveMap[splitKeypath[0]]; found {
+			subkeys := make(map[string]subkeyCfg)
+			for valueName, regKeyCfg := range regKeyConfig.Subkeys {
+				if len(regKeyCfg.Name) == 0 {
+					log.Warnf("the subkey %s of %s does not have a metric name, skipping", valueName, regKey)
+				} else {
+					subkeys[valueName] = regKeyCfg
+				}
+			}
+			c.registryKeys = append(c.registryKeys, registryKey{
+				name:            regKeyConfig.Name,
+				hive:            hive,
+				originalKeyPath: regKey,
+				keyPath:         strings.Join(splitKeypath[1:], "\\"),
+				subkeys:         subkeys,
+			})
+		} else {
+			return log.Errorf("unknown hive %s", splitKeypath[0])
+		}
+	}
+
+	return nil
+}
+
+// Run runs the check
+func (c *WindowsRegistryCheck) Run() error {
+	sender, err := c.GetSender()
+	if err != nil {
+		return err
+	}
+
+	for _, regKeyCfg := range c.registryKeys {
+		regKey, err := registry.OpenKey(regKeyCfg.hive, regKeyCfg.keyPath, registry.QUERY_VALUE)
+		if err != nil {
+			if errors.Is(err, fs.ErrPermission) {
+				// Treat access denied as errors
+				log.Errorf("access denied while accessing key %s: %s", regKeyCfg.originalKeyPath, err)
+			} else if errors.Is(err, registry.ErrNotExist) {
+				log.Warnf("key %s was not found: %s", regKeyCfg.originalKeyPath, err)
+			}
+		} else {
+			// if err == nil the key was opened, so we need to close it after we are done.
+			processRegistryKeyMetrics(sender, regKey, regKeyCfg)
+			regKey.Close()
+		}
+	}
+	sender.Commit()
+	return nil
+}
+
+func processRegistryKeyMetrics(sender sender.Sender, regKey registry.Key, regKeyCfg registryKey) {
+	for valueName, subkey := range regKeyCfg.subkeys {
+		_, valueType, err := regKey.GetValue(valueName, nil)
+		gaugeName := fmt.Sprintf("%s.%s.%s", checkPrefix, regKeyCfg.name, subkey.Name)
+		if errors.Is(err, registry.ErrNotExist) {
+			log.Warnf("value %s of key %s was not found: %s", valueName, regKeyCfg.name, err)
+			trySendDefaultValue(sender, subkey, gaugeName)
+		} else if errors.Is(err, fs.ErrPermission) {
+			log.Errorf("access denied while accessing value %s of key %s: %s", valueName, regKeyCfg.originalKeyPath, err)
+			trySendDefaultValue(sender, subkey, gaugeName)
```

Comment:
I wonder if we should add a tag specifying one of the possible reasons why we cannot send actual value (there are a few here).

Reply:
Yes that what I meant. Fair enough.

Reply:
yes, exactly what I meant bounded values like you mentioned. We control what we put and there are seems to be as many as calls to trySendDefaultValue (at most)